docs/dev/pagination-v2.md

```diff
@@ -0,0 +1,173 @@
+<!-- What applications will use pagination? -->
+
+Currently, SQL plugin does not provide SQL cursor syntax. However, the SQL REST endpoint can return result a page at a time.
+
+JDBC and ODBC drivers are the main users of cursor REST API.
+
+<!-- How will applications use pagination? -->
+
+# Scope
+<!-- What scenarios are excluded from this discussion? Why? -->
+
+The primary goal is to reach feature parity with the v1 engine. v1 engine suppors pagination only for simple `SELECT` queries with optional `WHERE` or `ORDER BY` clauses.
+
+## Compatibility with v1 engine
+<!-- How can one tell that v2 engine pagination is compatible with v1? -->
+
+The REST API will remain the same. However, v1 cursors are a thin wrapper over OpenSearch's scroll API.
+This is not possible with v2 engine because it may perform some processing in-memory
+
+# Behaviour Specification
+## Default Scenario
+
+A client requests a paginated response, retrieves data a subset at a time, and closes the cursor.
+
+```mermaid
+flowchart LR
+    start[query, fetch_size > 0]
+    fetch_some[get page]
+    close[close cursor]
+    start --> fetch_some 
+    fetch_some --> |more data | fetch_some 
+    fetch_some --> close
+    
+```
+Each subset response also includes a possibly new value for `cursor_id`. Client must ensure
+to pass the correct cursor id with each page request.
+
+
+### Client submits request with non-existing cursor_id 
+
+A client will receive an error response if it sends a request with a cursor_id that SQL plugin does not maintain.
+
+This can happen when client requests a paginated response but the cursor times out before all the data is retrieved.
+
+Response will include `error` and `status` fields.
+
+## Unable to create a new cursor [new]
+SQL plugin will have a limit on number of open cursors.
+
+A client will receive an error response when it requests a paginated response and the plugin reached the limit of possible open cursors.
+
+# Implementation
+
+## REST API
+Initial query
+```
+POST /_plugins/_sql
+{
+    "query" : "...",
+    "fetch_size": N
+}
+```
+
+Response
+```
+{
+  "cursor": /* cursor_id */,
+  "datarows": [
+    // ...
+    ],
+  "schema" : [
+    // ...
+  ]
+}
+```
+
+Get page
+```
+POST /_plugins/_sql
+{
+  "cursor": "<cursor_id>"
+}
+```
+response is same as initial response.
+
+Close cursor
+```
+POST /_plugins/_sql/close
+{
+    "cursor" : "<cursor_id>"
+}
+```
+
+Response
+```json
+{
+    "succeeded": true
+}
+
+```
+v1 engine always returns sucess on close even if the cursor is no longer alive as long as it recognizes cursor_id as a valid cursor identifier.
+
+v1 engine responds with an error such this if cursor_id is not a valid cursor identifier.
+```json
+{
+    "error": {
+        // ...
+    },
+    "status": 400
+}
+```
+
+### Non-scroll request
+
+For some queries, SQL plugin can fetch all data from OpenSearch at once.
```

Comment:
What are the conditions to make this a 'non-scroll request' vs a 'scroll request'.  I thought the conditions were based on whether all the data was returned (ie no cursor is needed)?  But you seem to indicate this is based on whether the user sends the fetch_size in the query.

Reply:
Dependent on the request by the user.  Got it.  Thanks.